File: webmvp/src/lib/songSearchText.ts
import type { Section } from "@/lib/types";

/** Max chars per index entry search blob (fits ~2000 entries under Firestore 1 MiB/chunk). */
export const SONG_INDEX_SEARCH_TEXT_MAX = 512;

export function flattenSectionsLyrics(sections: Section[]): string[] {
  return sections.flatMap((section) => section.lines.map((line) => line.lyrics));
}

export function buildSongSearchText(input: {
  title: string;
  artist?: string;
  tags?: string[];
  sections?: Section[];
}): string {
  const lyricLines = flattenSectionsLyrics(input.sections ?? []);
  const raw = [input.title, input.artist ?? "", ...(input.tags ?? []), ...lyricLines]
    .join(" ")
    .toLowerCase()
    .replace(/\s+/g, " ")
    .trim();

  return raw.slice(0, SONG_INDEX_SEARCH_TEXT_MAX);
}

File: mobile/lib/domain/song_search_text.dart
import 'package:lf_chords/domain/constants.dart';

List<String> flattenSectionsLyrics(List<Map<String, dynamic>> sections) {
  final lines = <String>[];
  for (final section in sections) {
    final sectionLines = section['lines'];
    if (sectionLines is! List) {
      continue;
    }
    for (final line in sectionLines) {
      if (line is Map && line['lyrics'] is String) {
        lines.add(line['lyrics'] as String);
      }
    }
  }
  return lines;
}

String buildSongSearchText({
  required String title,
  String? artist,
  List<String>? tags,
  List<Map<String, dynamic>>? sections,
}) {
  final lyricLines = flattenSectionsLyrics(sections ?? []);
  final raw = [
    title,
    artist ?? '',
    ...(tags ?? []),
    ...lyricLines,
  ].join(' ').toLowerCase().replaceAll(RegExp(r'\s+'), ' ').trim();

  if (raw.length <= songIndexSearchTextMax) {
    return raw;
  }
  return raw.substring(0, songIndexSearchTextMax);
}
